boogie_ast/src/boogie_program/writer.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! A writer for Boogie programs.
+//! Generates a text Boogie program with the following format:
+//! ```ignore
+//! // Type declarations:
+//! <Type declaration 1>
+//! <Type declaration 2>
+//! ...
+//!
+//! // Constant declarations:
+//! <Const declaration 1>
+//! <Const declaration 2>
+//! ...
+//!
+//! // Variable declarations:
+//! var <var-name1>: <type1>;
+//! var <var-name2>: <type2>;
+//! ...
+//!
+//! // Axioms
+//! axiom <expr1>;
+//! axiom <expr2>;
+//! ...
+//!
+//! // Functions:
+//! function <function1-name>(<arg1>: <type1>, ...) returns (return-var-name: <return-type>)
+//! {
+//!   <body>
+//! }
+//! ...
+//!
+//! // Procedures:
+//! procedure <procedure1-name>(<arg1>: <type1>, ...) returns (return-var-name: <return-type>)
+//!   requires <pre-condition1>;
+//!   requires <pre-condition2>;
+//!   ...
+//!   ensures <post-condition1>;
+//!   ensures <post-condition2>;
+//!   ...
+//!   modifies <var1>, <var2>, ...;
+//! {
+//!   <body>
+//! }
+//! ...
+//!
+//! ```
+use crate::boogie_program::*;
+
+use std::io::Write;
+
+/// A writer for Boogie programs.
+struct Writer<'a, T: Write> {
+    writer: &'a mut T,
+    indentation: usize,
+}
+
+impl<'a, T: Write> Writer<'a, T> {
+    fn new(writer: &'a mut T) -> Self {
+        Self { writer, indentation: 0 }
+    }
+
+    fn newline(&mut self) -> std::io::Result<()> {
+        writeln!(self.writer)
+    }
+
+    fn increase_indent(&mut self) {
+        self.indentation += 2;
+    }
+
+    fn decrease_indent(&mut self) {
+        self.indentation -= 2;
+    }
+
+    fn indent(&mut self) -> std::io::Result<()> {
+        write!(self.writer, "{:width$}", "", width = self.indentation)
+    }
+}
+
+impl BoogieProgram {
+    pub fn write_to<T: Write>(&self, writer: &mut T) -> std::io::Result<()> {
+        let mut writer = Writer::new(writer);
+
+        if !self.type_declarations.is_empty() {
+            writeln!(writer.writer, "// Type declarations:")?;
```

Comment:
nit: you could make implement Write for Writer, and avoid a bunch of these `writer.writer`

Reply:
Good idea! Done.